src/layout/Popover.js

```diff
@@ -1,36 +1,230 @@
-import React from 'react'
+import F from 'futil-js'
+import _ from 'lodash/fp'
+import React, { useRef, useEffect, useState, Children } from 'react'
+import { createPortal } from 'react-dom'
+import { Observer } from 'mobx-react'
 import OutsideClickHandler from 'react-outside-click-handler'
-import { observer } from 'mobx-react'
-import * as F from 'futil-js'
-
-// Simple popover
-let Popover = observer(
-  ({ isOpen, children, style }) =>
-    F.view(isOpen) && (
-      <OutsideClickHandler onOutsideClick={F.off(isOpen)}>
-        <div
-          style={{
-            position: 'relative',
-          }}
-        >
-          <div
-            className="popover"
-            style={{
-              position: 'absolute',
-              zIndex: 100,
-              fontWeight: 'normal',
-              textAlign: 'left',
-              background: 'white',
-              border: '1px solid #ebebeb',
-              ...style,
-            }}
-          >
-            {children}
-          </div>
-        </div>
-      </OutsideClickHandler>
+import {
+  moveX,
+  moveY,
+  middleX,
+  middleY,
+  area,
+  intersection,
+  fromObject,
+} from '../utils/rectangle'
+import { addEventListener, setNodeStyles } from '../utils/dom'
+
+let { min, max, floor } = Math
+
+let getPopoverTopleft = ({ windowRect, anchorRect, popoverRect, position }) => {
+  /*
+   * These rectangles represent the places we can render a popover, relative to
+   * its anchor element. They are positioned by applying translation functions:
+   * - The first two translations move a rectangle so that one of its boundaries
+   *   coincides with one of the boundaries of the anchor element.
+   * - The next two move a rectangle such that its visible area is maximized,
+   *   prioritizing the visibility of its topleft area.
+   *
+   * NOTE: The -1 in one of the translation functions is a hack to avoid text
+   * wrapping from making the popover content overflow into the anchor element.
+   */
+  let positioned = F.mapIndexed(
+    (translate, position) => ({ position, rectangle: translate(popoverRect) }),
+    {
+      top: _.flow(
```

Comment:
I tried to refactor it but I didn't think it worth it because it ended up being harder to read. Here it is for comparison purposes:

```javascript
let opposites = { top: 'bottom', bottom: 'top', left: 'right', right: 'left' }

move = (position) => {
  let isY = _.contains(position, ['top', 'bottom'])
  let [move, moveOpposite, middleFn] = isY
    ? [moveY, moveX, middleX] : 
    : [moveX, moveY, middleY]
  let adjust = isY ? 'right' : 'bottom'
  return _.flow(
    pop => move(anchorRef[position] - pop[opposites[position]])(pop),
    pop => moveOpposite(middleFn(anchorRect) - middleFn(pop))(pop),
    pop => moveOpposite(min(0, windowRect[adjust] - pop[adjust]))(pop),
    pop => moveOpposite(max(0, windowRect[opposites[adjust]] - pop[opposites[adjust]]))(pop)
  )
}
```